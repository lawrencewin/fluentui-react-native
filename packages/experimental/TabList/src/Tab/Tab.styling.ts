import React from 'react';

import type { IViewProps } from '@fluentui-react-native/adapters';
import { borderStyles, fontStyles } from '@fluentui-react-native/framework';
import type { Theme } from '@fluentui-react-native/framework';
import type { IconPropsV1 as IconProps } from '@fluentui-react-native/icon';
import type { LayoutEvent, PressablePropsExtended } from '@fluentui-react-native/interactive-hooks';
import type { TextProps } from '@fluentui-react-native/text';

import type { TabProps, TabSlotProps, TabTokens } from './Tab.types';
import type { TabListContextData } from '../TabList/TabList.types';

export const useTabSlotProps = (props: TabProps, tokens: TabTokens, theme: Theme, context: TabListContextData): TabSlotProps => {
  const { animatedIndicatorState, vertical } = context;

  const indicatorColor = React.useMemo(() => {
    if (props.tabKey === context.selectedKey) {
      // if we're the selected tab, we don't want to show the static indicator. We want to style the animated indicator with the correct color instead.
      return theme.colors.transparentStroke;
    } else {
      return tokens.indicatorColor;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [tokens.indicatorColor, props.tabKey, context.selectedKey, theme.colors.transparentStroke]);

  React.useEffect(() => {
    if (props.tabKey === context.selectedKey) {
      context.animatedIndicatorState.updateStyles({ indicator: { backgroundColor: tokens.indicatorColor } });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.tabKey, context.selectedKey, tokens.indicatorColor]);

  // onLayout callbacks to help calculate positioning of the animated indicator
  const onTabLayout = React.useCallback(
    (e: LayoutEvent) => {
      if (e.nativeEvent.layout) {
        // save x and y of tab
        animatedIndicatorState.addToLayoutMap(props.tabKey, {
          x: e.nativeEvent.layout.x,
          y: e.nativeEvent.layout.y,
          tabBorderWidth: tokens.borderWidth,
        });
      }
    },
    [animatedIndicatorState, props.tabKey, tokens.borderWidth],
  );

  const onIndicatorLayout = React.useCallback(
    (e: LayoutEvent) => {
      if (e?.nativeEvent?.layout) {
        // save width, height, and start margin of indicator.
        animatedIndicatorState?.addToLayoutMap(props.tabKey, {
          width: e.nativeEvent.layout.width,
          height: e.nativeEvent.layout.height,
          startMargin: tokens.indicatorInset,
        });
      }
    },
    [animatedIndicatorState, props.tabKey, tokens.indicatorInset],
  );

  const root = React.useMemo<PressablePropsExtended>(
    () => ({
      style: {
        display: 'flex',
        alignItems: 'center',
        flexDirection: tokens.flexDirection,
        alignSelf: 'flex-start',
        justifyContent: 'center',
        padding: 1,
        backgroundColor: tokens.backgroundColor,
        ...borderStyles.from(tokens, theme),
      },
      onLayout: onTabLayout,
    }),
    [tokens, theme, onTabLayout],
  );

  const contentContainer = React.useMemo<IViewProps>(
    () => ({
      style: {
        flexDirection: 'row',
        paddingStart: tokens.contentMarginStart,
        paddingEnd: tokens.contentMarginEnd,
      },
    }),
    [tokens.contentMarginStart, tokens.contentMarginEnd],
  );

  const content = React.useMemo<TextProps>(
    () => ({
      style: {
        color: tokens.color,
        ...fontStyles.from(tokens, theme),
      },
    }),
    [tokens, theme],
  );

  const icon = React.useMemo<IconProps>(() => {
    if (props.icon) {
      return { ...props.icon, color: tokens.iconColor, size: tokens.iconSize };
    }
    return {};
  }, [props.icon, tokens.iconColor, tokens.iconSize]);

  const stack = React.useMemo<IViewProps>(
    () => ({
      style: {
        display: 'flex',
        alignItems: 'center',
        flexDirection: 'row',
        alignSelf: 'flex-start',
        justifyContent: 'center',
        marginHorizontal: tokens.stackMarginHorizontal,
        marginVertical: tokens.stackMarginVertical,
      },
    }),
    [tokens.stackMarginHorizontal, tokens.stackMarginVertical],
  );

  const indicatorContainer = React.useMemo<IViewProps>(
    () => ({
      style: {
        backgroundColor: theme.colors.transparentBackground,
        ...(vertical
          ? {
              height: '100%',
              width: tokens.indicatorThickness,
              paddingVertical: tokens.indicatorInset,
            }
          : {
              width: '100%',
              height: tokens.indicatorThickness,
              paddingHorizontal: tokens.indicatorInset,
            }),
      },
    }),
    [vertical, tokens.indicatorThickness, tokens.indicatorInset, theme],
  );

  const indicator = React.useMemo<IViewProps>(
    () => ({
      style: {
        flex: 1,
        borderRadius: 99,
        backgroundColor: indicatorColor,
      },
      onLayout: onIndicatorLayout,
    }),
    [indicatorColor, onIndicatorLayout],
  );

  return { root, contentContainer, content, icon, stack, indicatorContainer, indicator };
};
