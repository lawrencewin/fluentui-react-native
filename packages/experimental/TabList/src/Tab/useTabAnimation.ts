import React from 'react';

import type { LayoutEvent } from '@fluentui-react-native/interactive-hooks';

import type { TabProps, TabTokens } from './Tab.types';
import type { TabListContextData } from '../TabList/TabList.types';

/**
 * On win32, when a vertical tablist initially lays out, we sometimes get odd, large height values that cause the
 * indicator to noticably take up the entire screen height before getting a correct layout value that makes sense. This is
 * an arbitrary limit we'll set to keep the indicator from looking weird.
 */
const RENDERING_HEIGHT_LIMIT = 20_000;

/**
 * This hook handles the logic on the tab side to correctly style and animate the TabListAnimatedIndicator.
 *
 * We save the layout information (width, height, x, y) of the Tab component using a LayoutEventHandler attached to the
 * root slot we initialize here, and we color the animated indicator using the user defined tab indicator color token.
 */
export function useTabAnimation(props: TabProps, context: TabListContextData, tokens: TabTokens) {
  const { animatedIndicatorState, selectedKey, vertical } = context;
  const { tabKey } = props;

  // If we're the selected tab, we style the TabListAnimatedIndicator with the correct token value set by the user
  React.useEffect(() => {
    if (tabKey === selectedKey) {
      animatedIndicatorState?.updateStyles({ indicator: { backgroundColor: tokens.indicatorColor } });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [tabKey, selectedKey, tokens.indicatorColor]);

  // onLayout callbacks to help calculate positioning of the animated indicator
  const onTabLayout = React.useCallback(
    (e: LayoutEvent) => {
      /**
       * This checks to see if we have relevant info to calculate the layout position and dimensions of the indicator. If this check fails, we don't
       * want to trigger a re-render by needlessly updating the TabList state.
       *
       * We also check if the info is good. Info can be bad in some weird cases:
       * - Check if width > 0 because there is an on-going issue caused by ScrollViews initially laying out its childrens' width to 0 and height to be a bigger than expected value.
       * - ScrollView also negatively affects the initial height values. For vertical TabLists, the initial height value will lay out incorrectly. Sometimes, the styling of the parent
       *   component combined with the ScrollView issues causes the initial height layout value to be completely unreasonable. Exactly which style that causes this issue isn't known;
       *   more investigation has to be done.
       */
      console.log(tabKey, e.nativeEvent.layout);
      if (
        e?.nativeEvent?.layout &&
        animatedIndicatorState?.tablistLayout &&
        animatedIndicatorState.tablistLayout.width > 0 &&
        e.nativeEvent.layout.height <= animatedIndicatorState.tablistLayout.height &&
        e.nativeEvent.layout.height < RENDERING_HEIGHT_LIMIT
      ) {
        let width: number, height: number;
        // Total Indicator inset consists of the horizontal/vertical margin of the indicator, the space taken up by the tab's focus border, and the
        // existing padding between the focus border and the tab itself. Multiply by 2 to account for the start + end margin/border/padding.
        const focusBorderPadding = 1;
        const totalIndicatorInset = 2 * (tokens.indicatorMargin + tokens.borderWidth + focusBorderPadding);
        // we can calculate the dimensions of the indicator using token values we have access to.
        if (vertical) {
          width = tokens.indicatorThickness;
          height = e.nativeEvent.layout.height - totalIndicatorInset;
        } else {
          width = e.nativeEvent.layout.width - totalIndicatorInset;
          height = tokens.indicatorThickness;
        }
        animatedIndicatorState?.addToLayoutMap(tabKey, {
          x: e.nativeEvent.layout.x,
          y: e.nativeEvent.layout.y,
          width: width,
          height: height,
          tabBorderWidth: tokens.borderWidth,
          startMargin: tokens.indicatorMargin,
        });
      }
    },
    [animatedIndicatorState, tabKey, tokens, vertical],
  );

  return onTabLayout;
}
