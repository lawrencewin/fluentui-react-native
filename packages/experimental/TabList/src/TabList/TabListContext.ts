import * as React from 'react';

import type { TabListContextData } from './TabList.types';

const nullFunction = () => null;

export const TabListContext = React.createContext<TabListContextData>({
  addTabKey: nullFunction,
  appearance: 'transparent',
  disabled: false,
  onTabSelect: nullFunction,
  removeTabKey: nullFunction,
  selectedKey: '',
  setSelectedTabRef: nullFunction,
  size: 'small',
  tabKeys: [],
  vertical: false,
});
