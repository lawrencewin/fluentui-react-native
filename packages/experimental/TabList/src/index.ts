export { tabName as tabsItemName } from './Tab.types';
export type { TabType, TabInfo, TabProps, TabSlotProps, TabState, TabTokens } from './Tab.types';
export { Tab as TabsItem } from './Tab';
export { tabListName as tabsName } from './TabList.types';
export type {
  TabListContextData,
  TabListInfo,
  TabListProps,
  TabListSlotProps,
  TabListState,
  TabListTokens,
  TabListType,
} from './TabList.types';
export { TabList as Tabs, TabListContext as TabsContext } from './TabList';
