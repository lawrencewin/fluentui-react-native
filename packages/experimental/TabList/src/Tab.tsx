/** @jsx withSlots */
import * as React from 'react';
import { Pressable, View } from 'react-native';

import type { UseSlots } from '@fluentui-react-native/framework';
import { compose, mergeProps, withSlots } from '@fluentui-react-native/framework';
import { Icon, createIconProps } from '@fluentui-react-native/icon';
import { TextV1 as Text } from '@fluentui-react-native/text';

import { stylingSettings } from './Tab.styling';
import type { TabType, TabProps } from './Tab.types';
import { tabName } from './Tab.types';
import { TabListContext } from './TabList';
import { useTab } from './useTab';

export const Tab = compose<TabType>({
  displayName: tabName,
  ...stylingSettings,
  slots: {
    root: Pressable,
    stack: View,
    icon: Icon,
    indicator: View,
    content: Text,
  },
  useRender: (userProps: TabProps, useSlots: UseSlots<TabType>) => {
    const tabsItem = useTab(userProps);

    const iconProps = createIconProps(userProps.icon);
    const context = React.useContext(TabListContext);

    // Grab the styled slots.
    const Slots = useSlots(userProps, (layer) => tabsItem.state[layer] || userProps[layer]);
    // Return the handler to finish render.
    return (final: TabProps, ...children: React.ReactNode[]) => {
      if (!tabsItem.state) {
        return null;
      }

      const { icon, itemKey, itemCount, headerText, ...mergedProps } = mergeProps(tabsItem.props, final);

      let containerText = headerText;
      if (itemCount !== undefined) {
        containerText += ` (${itemCount})`;
      }

      const renderContent = !!headerText || itemCount !== undefined;
      context?.views?.set(itemKey, children);

      return (
        <Slots.root {...mergedProps}>
          <Slots.stack>
            {icon && <Slots.icon {...iconProps} />}
            {renderContent && <Slots.content key="content">{containerText}</Slots.content>}
          </Slots.stack>
          <Slots.indicator />
        </Slots.root>
      );
    };
  },
});

export default Tab;
