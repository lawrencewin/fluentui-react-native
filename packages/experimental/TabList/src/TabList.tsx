/** @jsx withSlots */
import * as React from 'react';
import { Pressable, View } from 'react-native';

import { FocusZone } from '@fluentui-react-native/focus-zone';
import type { UseSlots } from '@fluentui-react-native/framework';
import { compose, mergeProps, withSlots } from '@fluentui-react-native/framework';
import { TextV1 as Text } from '@fluentui-react-native/text';

import { stylingSettings } from './TabList.styling';
import type { TabListType, TabListProps, TabListContextData } from './TabList.types';
import { tabListName } from './TabList.types';
import { useTabList } from './useTabList';

export const TabListContext = React.createContext<TabListContextData>({
  selectedKey: null,
  onTabsClick: (/* key: string */) => {
    return;
  },
  getTabId: (/* key:string, index: number*/) => {
    return null;
  },
  updateSelectedTabsItemRef: (/* ref: React.RefObject<any>*/) => {
    return;
  },
  tabsItemKeys: [],
  views: null,
});

export const TabList = compose<TabListType>({
  displayName: tabListName,
  ...stylingSettings,
  slots: {
    root: Pressable,
    label: Text,
    container: FocusZone,
    stack: View,
    tabPanel: View,
  },
  useRender: (userProps: TabListProps, useSlots: UseSlots<TabListType>) => {
    // configure props and state for tabs based on user props
    const tabs = useTabList(userProps);

    // Grab the styled slots.
    const Slots = useSlots(userProps, (layer) => tabs.state[layer] || userProps[layer]);

    // Return the handler to finish render.
    return (final: TabListProps, ...children: React.ReactNode[]) => {
      if (!tabs.state) {
        return null;
      }

      const { label, defaultTabbableElement, isCircularNavigation, ...mergedProps } = mergeProps(tabs.props, final);

      // Populate the tabsItemKeys array.
      if (children) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore - TODO, fix typing error
        tabs.state.context.tabsItemKeys = React.Children.map(children, (child: React.ReactChild) => {
          if (React.isValidElement(child)) {
            // Sets default selected tabItem.
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore - TODO, fix typing error
            if (tabs.state?.context.selectedKey == null && !child.props.disabled) {
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore - TODO, fix typing error
              tabs.state.context.selectedKey = child.props.itemKey;
            }
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore - TODO, fix typing error
            return child.props.itemKey;
          }
        });
      }

      return (
        <TabListContext.Provider
          // Passes in the selected key and a hook function to update the newly selected tabsItem and call the client's onTabsClick callback.
          value={tabs.state.context}
        >
          <Slots.root {...mergedProps}>
            {tabs?.state?.label && <Slots.label key="label">{label}</Slots.label>}
            <Slots.container defaultTabbableElement={defaultTabbableElement} isCircularNavigation={isCircularNavigation}>
              <Slots.stack>{children}</Slots.stack>
            </Slots.container>
            <Slots.tabPanel>
              <TabListContext.Consumer>
                {(context) => !tabs?.state?.headersOnly && <View>{context.views.get(context.selectedKey)}</View>}
              </TabListContext.Consumer>
            </Slots.tabPanel>
          </Slots.root>
        </TabListContext.Provider>
      );
    };
  },
});

export default TabList;
