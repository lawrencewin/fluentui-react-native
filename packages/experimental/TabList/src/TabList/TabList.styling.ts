import type { Theme, UseStylingOptions } from '@fluentui-react-native/framework';
import { buildProps, layoutStyles } from '@fluentui-react-native/framework';

import { tabListName } from './TabList.types';
import type { TabListTokens, TabListSlotProps, TabListProps } from './TabList.types';
import { defaultTabListTokens } from './TabListTokens';

export const stylingSettings: UseStylingOptions<TabListProps, TabListSlotProps, TabListTokens> = {
  tokens: [defaultTabListTokens, tabListName],
  states: ['vertical'],
  slotProps: {
    stack: buildProps(
      (tokens: TabListTokens) => ({
        style: {
          flexDirection: tokens.direction,
        },
      }),
      ['direction'],
    ),
    root: buildProps(
      (tokens: TabListTokens, theme: Theme) => ({
        style: layoutStyles.from(tokens, theme),
      }),
      layoutStyles.keys,
    ),
  },
};
