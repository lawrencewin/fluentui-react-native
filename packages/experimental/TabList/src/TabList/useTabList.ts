import * as React from 'react';
import type { View, AccessibilityState, ViewStyle } from 'react-native';

import { memoize } from '@fluentui-react-native/framework';
import { useSelectedKey } from '@fluentui-react-native/interactive-hooks';

import type { AnimatedIndicatorState, ListLayoutInfo, TabLayoutInfo } from './TabList.types';
import { type TabListProps, type TabListInfo, type TabListState } from './TabList.types';

/**
 * Re-usable hook for TabList.
 * This hook configures props and state for TabList.
 *
 * @param props user props sent to TabList
 * @returns configured props and state for TabList
 */
export const useTabList = (props: TabListProps): TabListInfo => {
  const defaultComponentRef = React.useRef(null);
  const {
    accessible,
    appearance = 'transparent',
    accessibilityState,
    componentRef = defaultComponentRef,
    defaultSelectedKey,
    disabled = false,
    isCircularNavigation,
    onTabSelect,
    selectedKey,
    size = 'medium',
    vertical = false,
  } = props;

  const data = useSelectedKey(selectedKey || defaultSelectedKey || null, onTabSelect);

  // selectedTabRef should be set to default tabbable element.
  const [selectedTabRef, setSelectedTabRef] = React.useState(React.useRef<View>(null));
  const [invoked, setInvoked] = React.useState(false);
  const [tabKeys, setTabKeys] = React.useState<string[]>([]);
  const [listLayoutInfo, setListLayoutInfo] = React.useState<ListLayoutInfo>({});
  const [userDefinedAnimatedIndicatorStyles, setUserDefinedAnimatedIndicatorStyles] = React.useState<
    Partial<AnimatedIndicatorState['styles']>
  >({});

  const addTabKey = React.useCallback(
    (tabKey: string) => {
      if (__DEV__ && tabKeys.includes(tabKey)) {
        console.warn(`Tab Key "${tabKey}" already exists in the TabList. Duplicate keys are not supported.`);
      }
      setTabKeys((keys) => [...keys, tabKey]);
    },
    [tabKeys, setTabKeys],
  );

  const removeTabKey = React.useCallback(
    (tabKey: string) => {
      setTabKeys((keys) => keys.filter((key) => key !== tabKey));
    },
    [setTabKeys],
  );

  const addToLayoutMap = React.useCallback(
    (tabKey: string, layoutInfo: TabLayoutInfo) => {
      setListLayoutInfo((prev) => ({ ...prev, [tabKey]: { ...prev[tabKey], ...layoutInfo } }));
    },
    [setListLayoutInfo],
  );

  const selectedIndicatorLayout = React.useMemo<TabLayoutInfo | null>(() => {
    const key = selectedKey ?? data.selectedKey;
    return key ? listLayoutInfo[key] : null;
  }, [selectedKey, data.selectedKey, listLayoutInfo]);

  const animatedIndicatorStyles = React.useMemo<AnimatedIndicatorState['styles']>(() => {
    // if not all layout props have been recorded for the current selected indicator, don't render the animated indicator
    if (selectedIndicatorLayout) {
      const { x, y, width, height, startMargin, tabBorderWidth } = selectedIndicatorLayout;
      if (
        x !== undefined &&
        y !== undefined &&
        width !== undefined &&
        height !== undefined &&
        startMargin !== undefined &&
        tabBorderWidth !== undefined
      ) {
        // calculate styles here
        const containerStyles: ViewStyle = { ...userDefinedAnimatedIndicatorStyles.container };
        const indicatorStyles: ViewStyle = {
          ...userDefinedAnimatedIndicatorStyles.indicator,
          width: selectedIndicatorLayout.width,
          height: selectedIndicatorLayout.height,
        };
        if (vertical) {
          containerStyles.start = tabBorderWidth + 1;
          indicatorStyles.marginTop = selectedIndicatorLayout.y + selectedIndicatorLayout.startMargin + tabBorderWidth + 1;
        } else {
          containerStyles.bottom = height + y + 1;
          indicatorStyles.marginLeft = selectedIndicatorLayout.x + selectedIndicatorLayout.startMargin + tabBorderWidth + 1;
        }
        return {
          container: containerStyles,
          indicator: indicatorStyles,
        };
      }
    }
    return {
      container: { display: 'none' },
      indicator: { display: 'none' },
    };
  }, [vertical, selectedIndicatorLayout, userDefinedAnimatedIndicatorStyles]);

  const state: TabListState = {
    context: {
      addTabKey: addTabKey,
      animatedIndicatorState: {
        addToLayoutMap: addToLayoutMap,
        layout: listLayoutInfo,
        styles: animatedIndicatorStyles,
        updateStyles: setUserDefinedAnimatedIndicatorStyles,
      },
      appearance: appearance,
      disabled: disabled,
      invoked: invoked,
      onTabSelect: data.onKeySelect,
      removeTabKey: removeTabKey,
      selectedKey: selectedKey ?? data.selectedKey,
      setSelectedTabRef: setSelectedTabRef,
      setInvoked: setInvoked,
      size: size,
      tabKeys: tabKeys,
      vertical: vertical,
    },
  };

  return {
    props: {
      ...props,
      accessible: accessible ?? true,
      accessibilityState: getAccessibilityState(disabled, accessibilityState),
      accessibilityRole: 'tablist',
      appearance: appearance,
      componentRef: componentRef,
      defaultTabbableElement: selectedTabRef,
      isCircularNavigation: isCircularNavigation ?? false,
      size: size,
      vertical: vertical,
    },
    state: { ...state },
  };
};

const getAccessibilityState = memoize(getAccessibilityStateWorker);
function getAccessibilityStateWorker(disabled: boolean, accessibilityState?: AccessibilityState): AccessibilityState {
  if (accessibilityState) {
    return { disabled, ...accessibilityState };
  }
  return { disabled };
}
