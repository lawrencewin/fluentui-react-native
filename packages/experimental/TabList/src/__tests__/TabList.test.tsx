import * as React from 'react';
// import type { I18nManagerStatic } from 'react-native';
// import { I18nManager } from 'react-native';

import { checkRenderConsistency, checkReRender } from '@fluentui-react-native/test-tools';
import * as renderer from 'react-test-renderer';

import Tab from '../Tab/Tab';
import TabList from '../TabList/TabList';

describe('TabList component tests', () => {
  beforeAll(() => {
    jest.mock('react-native/Libraries/Utilities/Platform', () => ({
      OS: 'win32',
      select: () => null,
    }));
    jest.mock('react-native/Libraries/ReactNative/I18nManager', () => ({
      isRtl: false,
    }));
  });

  it('TabList default props', () => {
    const tree = renderer
      .create(
        <TabList>
          <Tab tabKey="1">Tab 1</Tab>
          {/* <Tab tabKey="2">Tab 2</Tab>
          <Tab tabKey="3">Tab 3</Tab> */}
        </TabList>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it('TabList disabled tabs', () => {
  //   const tree = renderer
  //     .create(
  //       <TabList>
  //         <Tab disabled tabKey="1">
  //           Tab 1
  //         </Tab>
  //         <Tab tabKey="2">Tab 2</Tab>
  //         <Tab disabled tabKey="3">
  //           Tab 3
  //         </Tab>
  //       </TabList>,
  //     )
  //     .toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  // it('TabList disabled list', () => {
  //   const tree = renderer
  //     .create(
  //       <TabList disabled>
  //         <Tab tabKey="1">Tab 1</Tab>
  //         <Tab tabKey="2">Tab 2</Tab>
  //         <Tab tabKey="3">Tab 3</Tab>
  //       </TabList>,
  //     )
  //     .toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  // it('TabList appearance', () => {
  //   const tree = renderer
  //     .create(
  //       <TabList appearance="subtle">
  //         <Tab tabKey="1">Tab 1</Tab>
  //         <Tab tabKey="2">Tab 2</Tab>
  //         <Tab tabKey="3">Tab 3</Tab>
  //       </TabList>,
  //     )
  //     .toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  // it('TabList size', () => {
  //   const tree = renderer
  //     .create(
  //       <TabList size="large">
  //         <Tab tabKey="1">Tab 1</Tab>
  //         <Tab tabKey="2">Tab 2</Tab>
  //         <Tab tabKey="3">Tab 3</Tab>
  //       </TabList>,
  //     )
  //     .toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  // it('TabList orientation', () => {
  //   const tree = renderer
  //     .create(
  //       <TabList vertical>
  //         <Tab tabKey="1">Tab 1</Tab>
  //         <Tab tabKey="2">Tab 2</Tab>
  //         <Tab tabKey="3">Tab 3</Tab>
  //       </TabList>,
  //     )
  //     .toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  // it('TabList rendering icons', () => {
  //   const tree = renderer
  //     .create(
  //       <TabList>
  //         <Tab
  //           icon={{
  //             fontSource: {
  //               fontFamily: 'Arial',
  //               codepoint: 0x2663,
  //             },
  //           }}
  //           tabKey="1"
  //         >
  //           Tab 1
  //         </Tab>
  //         <Tab
  //           icon={{
  //             fontSource: {
  //               fontFamily: 'Arial',
  //               codepoint: 0x2663,
  //             },
  //           }}
  //           tabKey="2"
  //         />
  //       </TabList>,
  //     )
  //     .toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  // it('TabList simple rendering does not invalidate styling', () => {
  //   checkRenderConsistency(
  //     () => (
  //       <TabList>
  //         <Tab tabKey="1">Tab 1</Tab>
  //         <Tab tabKey="2">Tab 2</Tab>
  //         <Tab tabKey="3">Tab 3</Tab>
  //       </TabList>
  //     ),
  //     2,
  //   );
  // });

  // it('TabList re-renders correctly', () => {
  //   checkReRender(
  //     () => (
  //       <TabList>
  //         <Tab tabKey="1">Tab 1</Tab>
  //         <Tab tabKey="2">Tab 2</Tab>
  //         <Tab tabKey="3">Tab 3</Tab>
  //       </TabList>
  //     ),
  //     2,
  //   );
  // });
  afterAll(() => {
    jest.unmock('react-native/Libraries/Utilities/Platform');
    jest.unmock('react-native/Libraries/ReactNative/I18nManager');
  });
});
