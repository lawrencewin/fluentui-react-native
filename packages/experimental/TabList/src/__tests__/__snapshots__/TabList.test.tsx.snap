// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList component tests TabList default props 1`] = `
<RCTFocusZone
  focusZoneDirection="horizontal"
  navigateAtEnd="NavigateStopAtEnds"
>
  <View
    accessibilityRole="tablist"
    accessibilityState={
      {
        "disabled": false,
      }
    }
    accessible={true}
    appearance="transparent"
    componentRef={
      {
        "current": null,
      }
    }
    size="medium"
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityActions={
        [
          {
            "name": "Select",
          },
        ]
      }
      accessibilityLabel="Tab 1"
      accessibilityPositionInSet={0}
      accessibilityRole="tab"
      accessibilitySetSize={0}
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onAccessibilityAction={[Function]}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "#00000000",
          "borderColor": "#00000000",
          "borderRadius": 4,
          "borderWidth": 2,
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
          "padding": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "alignSelf": "flex-start",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 12,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "paddingEnd": 2,
              "paddingStart": 2,
            }
          }
        >
          <Text
            accessible={false}
            ellipsizeMode="tail"
            numberOfLines={0}
            onAccessibilityTap={[Function]}
            style={
              {
                "color": "#424242",
                "fontFamily": "Segoe UI",
                "fontSize": 12,
                "fontWeight": "400",
                "margin": 0,
              }
            }
          >
            Tab 1
          </Text>
        </View>
      </View>
      <View
        color="#00000000"
        inset={12}
        style={
          {
            "backgroundColor": "#00000000",
            "height": 2,
            "paddingHorizontal": 12,
            "width": "100%",
          }
        }
        thickness={2}
        vertical={false}
      >
        <View
          style={
            {
              "backgroundColor": "#00000000",
              "borderRadius": 99,
              "flex": 1,
            }
          }
        />
      </View>
    </View>
  </View>
</RCTFocusZone>
`;
